newsboat.cpp

```diff
@@ -188,7 +207,22 @@ int main(int argc, char* argv[])
 	CliArgsParser args(argc, argv);
 
 	if (args.should_print_usage()) {
-		print_usage(args.program_name());
+		std::string config_path = utils::resolve_tilde("~/.newsboat/config");
+		if (args.set_config_file()) {
+			config_path = args.config_file();
+		}
```

Comment:
I expect this does not result in the actual location after creating files in XGD locations (`~/.config/newsboat/xxx`) which might be confusing.

Same for `urls_path` and `cache_path`.

Reply:
Wow, that's embarrassing. My code actually makes no sense at all!

Thank you for the great catch :) I think I should move `ConfigPaths` out of the `Controller` and into `main()` to fix this; let me try that.

Reply:
Here's proof that with new commit, everything works as intended. I redacted the options to make it a bit shorter.

```
# I don't have XDG directories, so Newsboat uses dotdir
$ XDG_CONFIG_HOME= XDG_DATA_HOME= ./newsboat --help
Files:
	- config:  /home/minoru/.newsboat/config
	- urls:    /home/minoru/.newsboat/urls
	- cache:   /home/minoru/.newsboat/cache.db

$ HOME=/wonderland XDG_CONFIG_HOME= XDG_DATA_HOME= ./newsboat --help
Files:
	- config:  /wonderland/.newsboat/config
	- urls:    /wonderland/.newsboat/urls
	- cache:   /wonderland/.newsboat/cache.db

# /dev/shm is tmpfs
$ mkdir /dev/shm/newsboat-xdg-test

# For the following tests, note that Newsboat switches to XDG *only* if XDG
# config directory for Newsboat exist

# Neither config nor data dirs for Newsboat exist, so falling back to dotdir
$ HOME=/home/minoru XDG_CONFIG_HOME= XDG_DATA_HOME=/dev/shm/newsboat-xdg-test ./newsboat --help
Files:
	- config:  /home/minoru/.newsboat/config
	- urls:    /home/minoru/.newsboat/urls
	- cache:   /home/minoru/.newsboat/cache.db

# Still neither directory exists, so dotdir again
$ HOME=/home/minoru XDG_CONFIG_HOME=/dev/shm/newsboat-xdg-test XDG_DATA_HOME=/dev/shm/newsboat-xdg-test ./newsboat --help
Files:
	- config:  /dev/shm/newsboat-xdg-test/newsboat/config
	- urls:    /dev/shm/newsboat-xdg-test/newsboat/urls
	- cache:   /dev/shm/newsboat-xdg-test/newsboat/cache.db

# Ditto
$ HOME=/home/minoru XDG_CONFIG_HOME=/dev/shm/newsboat-xdg-test XDG_DATA_HOME= ./newsboat --help
Files:
	- config:  /dev/shm/newsboat-xdg-test/newsboat/config
	- urls:    /dev/shm/newsboat-xdg-test/newsboat/urls
	- cache:   /home/minoru/.local/share/newsboat/cache.db

# Now we have a directory for Newsboat
$ mkdir /dev/shm/newsboat-xdg-test/newsboat

# XDG_CONFIG_HOME is not set, so defaults to ~/.config. The ~/.config/newsboat
# doesn't exist, so falling back to dotdir
$ HOME=/home/minoru XDG_CONFIG_HOME= XDG_DATA_HOME=/dev/shm/newsboat-xdg-test ./newsboat --help
Files:
	- config:  /home/minoru/.newsboat/config
	- urls:    /home/minoru/.newsboat/urls
	- cache:   /home/minoru/.newsboat/cache.db

# Finally found the config dir for Newsboat, so using XDG
$ HOME=/home/minoru XDG_CONFIG_HOME=/dev/shm/newsboat-xdg-test XDG_DATA_HOME=/dev/shm/newsboat-xdg-test ./newsboat --help
Files:
	- config:  /dev/shm/newsboat-xdg-test/newsboat/config
	- urls:    /dev/shm/newsboat-xdg-test/newsboat/urls
	- cache:   /dev/shm/newsboat-xdg-test/newsboat/cache.db

# Config dir for Newsboat found, so using XDG -- but XDG_DATA_HOME is not set, so default
$ HOME=/home/minoru XDG_CONFIG_HOME=/dev/shm/newsboat-xdg-test XDG_DATA_HOME= ./newsboat --help
Files:
	- config:  /dev/shm/newsboat-xdg-test/newsboat/config
	- urls:    /dev/shm/newsboat-xdg-test/newsboat/urls
	- cache:   /home/minoru/.local/share/newsboat/cache.db
```

Reply:
Nice solution, looks good to me.

Just to be sure, the comment `# Still neither directory exists, so dotdir again` shows XDG directories. Am I misunderstanding something there?

Reply:
My bad: I originally run the commands without redefining `HOME`, but then decided to do it in all commands for clarity. So I re-run the commands, but forgot to delete /dev/shm/newsboat-xdg-test beforehand. The proof is now amended. Thanks for being so attentive!